src/univariate/discrete/negativebinomial.jl

```diff
@@ -129,3 +129,38 @@ function cf(d::NegativeBinomial, t::Real)
     r, p = params(d)
     return (((1 - p) * cis(t)) / (1 - p * cis(t)))^r
 end
+
+# ChainRules definitions
+
+function ChainRulesCore.rrule(::typeof(logpdf), d::NegativeBinomial, k::Real)
+    # Compute log probability
+    r, p = Distributions.params(d)
+    edgecase = isone(p) && iszero(k)
+    insupp = insupport(d, k)
+    
+    # Primal computation
+    Ω = r * log(p) + k * log1p(-p)
+    if edgecase
+        Ω = zero(Ω)
+    elseif !insupp
+        Ω = oftype(Ω, -Inf)
+    else
+        Ω = Ω - log(k + r) - logbeta(r, k + 1)
+    end
+
+    # Define pullback
+    function logpdf_NegativeBinomial_pullback(Δ)
+        Δr = Δ * (log(p) - inv(k + r) - digamma(r) + digamma(r + k + 1))
+        Δp = Δ * (r / p - k * inv(1 - p))
+        Δd = if edgecase
+            ChainRulesCore.Tangent{typeof(d)}(; r=Δr, p=NaN)
+        elseif !insupp
+            ChainRulesCore.Tangent{typeof(d)}(; r=zero(Δr), p=zero(Δp))
+        else
+            ChainRulesCore.Tangent{typeof(d)}(; r=Δr, p=Δp)
+        end
```

Comment:
The `NaN` creates some (potential) type stability issues. I think it might be easier to just compute `Δr` and `Δp` in a type stable way, and then define `Δd = ChainRulesCore.Tangent{typeof(d)}(; r=Δr, p=Δp)`.

As I mentioned on Slack, the tangents for `!insupp` should be `oftype(..., NaN)`.

Reply:
Not sure I understand. Is this closer to what you mean?
```suggestion
function logpdf_NegativeBinomial_pullback(Δ)
    Δr = Δ * (log(p) - inv(k + r) - digamma(r) + digamma(r + k + 1))
    Δp = Δ * (r / p - k * inv(1 - p))
    if edgecase
        Δp = oftype(Δp, NaN)
    elseif !insupp
        Δr = oftype(Δr, NaN)
        Δp = oftype(Δp, NaN)
    end
    Δd = ChainRulesCore.Tangent{typeof(d)}(; r=Δr, p=Δp)
    return ChainRulesCore.NoTangent(), Δd, ChainRulesCore.NoTangent()
end
```

Reply:
Yes, that's closer to what I mean. I'm not completely sure if `NaN` is correct in the edge case though?

Reply:
Probably not. Let me write the tests first.